modules/wallet/scan.go

```diff
@@ -0,0 +1,198 @@
+		}
+		for _, txn := range block.Transactions {
+			for i, sco := range txn.SiacoinOutputs {
+				// if a seed output is found, add it to the output set
+				if index, exists := s.keys[sco.UnlockHash]; exists && sco.Value.Cmp(s.dustThreshold) > 0 {
+					id := types.OutputID(txn.SiacoinOutputID(uint64(i)))
+					s.siacoinOutputs[id] = scannedOutput{
+						id:        id,
+						value:     sco.Value,
+						seedIndex: index,
+					}
+				}
+			}
+			// TODO: should this be done in a separate txn loop? This would be
+			// mandatory if txn0 can spend an output created in txn1
+			for _, sci := range txn.SiacoinInputs {
+				// if a seed output is spent, remove it from the output set
+				id := types.OutputID(sci.ParentID)
+				if _, exists := s.siacoinOutputs[id]; exists {
+					delete(s.siacoinOutputs, id)
+				}
+			}
+		}
+	}
+
+	// update s.largestIndexSeen
+	var addrs []types.UnlockHash
+	for _, diff := range cc.SiacoinOutputDiffs {
+		addrs = append(addrs, diff.SiacoinOutput.UnlockHash)
+	}
+	for _, diff := range cc.SiafundOutputDiffs {
+		addrs = append(addrs, diff.SiafundOutput.UnlockHash)
+	}
+	for _, block := range cc.AppliedBlocks {
+		for _, mp := range block.MinerPayouts {
+			addrs = append(addrs, mp.UnlockHash)
+		}
+		for _, txn := range block.Transactions {
+			for _, sci := range txn.SiacoinInputs {
+				addrs = append(addrs, sci.UnlockConditions.UnlockHash())
+			}
+			for _, sco := range txn.SiacoinOutputs {
+				addrs = append(addrs, sco.UnlockHash)
+			}
+			for _, sfi := range txn.SiafundInputs {
+				addrs = append(addrs, sfi.UnlockConditions.UnlockHash())
+			}
+			for _, sfo := range txn.SiafundOutputs {
+				addrs = append(addrs, sfo.UnlockHash)
+			}
+		}
+	}
+	for _, addr := range addrs {
+		index, exists := s.keys[addr]
+		if exists && index > s.largestIndexSeen {
+			s.largestIndexSeen = index
+		}
+	}
+}
+
+// scan subscribes s to cs and scans the blockchain for addresses that belong
+// to s's seed. If scan returns errMaxKeys, additional keys may need to be
+// generated to find all the addresses.
+func (s *seedScanner) scan(cs modules.ConsensusSet) error {
+	// generate a bunch of keys and scan the blockchain looking for them. If
+	// none of the 'upper' half of the generated keys are found, we are done;
+	// otherwise, generate more keys and try again (bounded by a sane
+	// default).
+	//
+	// NOTE: since scanning is very slow, we aim to only scan once, which
+	// means generating many keys.
+	var numKeys uint64 = numInitialKeys
+	for s.numKeys() < maxScanKeys {
+		s.generateKeys(numKeys)
+		if err := cs.ConsensusSetSubscribe(s, modules.ConsensusChangeBeginning); err != nil {
+			return err
+		}
+		if s.largestIndexSeen < s.numKeys()/2 {
+			cs.Unsubscribe(s)
+			return nil
+		}
+		// quadruple number of keys generated each iteration, capping so that we
+		// do not exceed maxScanKeys
+		numKeys *= 4
+		if numKeys > maxScanKeys-s.numKeys() {
+			numKeys = maxScanKeys - s.numKeys()
+		}
+	}
+	cs.Unsubscribe(s)
```

Comment:
I am pretty confident that using a separate subscription from the wallet is going to cause headaches and missed outputs at least a few times for the users. The wallet and the scanner need to be on the same page.